### ISSUE TITLE
questions about the code

### ISSUE THREAD

**Issue by handsomelys** [maple-research-lab/CaCo#1]
```
class CaCo_PN(nn.Module):
    def __init__(self,bank_size,dim):
        super(CaCo_PN, self).__init__()
        self.register_buffer("W", torch.randn(dim, bank_size))
        self.register_buffer("v", torch.zeros(dim, bank_size))
    def forward(self,q):
        memory_bank = self.W
        memory_bank = nn.functional.normalize(memory_bank, dim=0)
        logit=torch.einsum('nc,ck->nk', [q, memory_bank])
        return memory_bank, self.W, logit
    def update(self, m, lr, weight_decay, g):
        g = g + weight_decay * self.W
        self.v = m * self.v + g
        self.W = self.W - lr * self.v
    def print_weight(self):
        print(torch. Sum(self.W).item())
```

Hello, I'm a beginner. This code confuses me a bit, what do` self.v` and `self.W` mean? How does the `update` function work?
Also,  I can't find the corresponding part of the paper, can you help me point it out?
 Looking forward to your reply!

**Comment by wang3702** [comment]
Sorry for the late response!
self.W is the weight of the memoery bank. self.v keeps track of the momentum. Here update just mimics the SGD optimizer. You can also achieve this by memory bank.
It's a mimic of optimizer, so you can not find in the paper.

**Comment by handsomelys** [comment]
Thanks for your reply! 
But I still have a question, which piece of the code corresponds to the formula (10) in the paper? I searched for a long time but wasn't sure.
And what is the difference between `update_multicrop_network` and `update_sym_network` in train_caco.py? Because I'm not a computer vision practitioner, these questions bother me a bit, but I'm interested in your work

**Comment by wang3702** [comment]
Sorry for my previous late response! It did not send the issue to my email! Thank you so much for your interest in my work.

https://github.com/maple-research-lab/CaCo/blob/main/training/train_caco.py#L184-L210 It's here.
update_multicrop_network is for multi-crop training with input include standard 224*224 crop as well as small 96*96 crop. update_sym_network is for standard 224*224 standard crop. Basically, you only need to check update_sym_network for general purposes. Also, Eq. (7) is enough to understand our gradient update rule, while Eq. (8) to Eq.(10) is for general discussion.

With any further questions, please don't hesitate to raise here.

**Comment by handsomelys** [comment]
Thank you for your reply. I wanted to migrate CaCo to my task, but I found it somewhat difficult. Because I found that I couldn't simply migrate CaCo in the same way I migrated MoCo to my task. Because the code of CaCo is a bit complicated, and there are some variables in it I don't quite understand its meaning, if you can, I hope you can add a little more comments in the code of CaCo in the future. LOL

**Comment by wang3702** [comment]
Sorry that my code raised difficulty for your project. I think you can simply modify this code to a function so that you can directly call in your project:
```
    with torch.no_grad():
        # update memory bank

        # logits: Nx(1+K)
        logits1 = logits_keep1/cur_adco_t#/args.mem_t
        # negative logits: NxK
        # logits: Nx(1+K)
        logits2 = logits_keep2/cur_adco_t#/args.mem_t
        
        p_qd1 = nn.functional.softmax(logits1, dim=1)
        p_qd1[torch.arange(logits1.shape[0]), filter_index1] = 1 - p_qd1[torch.arange(logits1.shape[0]), filter_index1]

        g1 = torch.einsum('cn,nk->ck', [q_pred.T, p_qd1]) / logits1.shape[0] - torch.mul(
            torch.mean(torch.mul(p_qd1, logits_keep1), dim=0), d_norm1)
        
        
        p_qd2 = nn.functional.softmax(logits2, dim=1)
        p_qd2[torch.arange(logits1.shape[0]), filter_index2] = 1 - p_qd2[torch.arange(logits2.shape[0]), filter_index2]



        g2 = torch.einsum('cn,nk->ck', [k_pred.T, p_qd2]) / logits2.shape[0] - torch.mul(
            torch.mean(torch.mul(p_qd2, logits_keep2), dim=0), d_norm2)
        g = -torch.div(g1, torch.norm(d1, dim=0))  - torch.div(g2,torch.norm(d2, dim=0))#/ args.mem_t  # c*k
        g /=cur_adco_t
        
        g = all_reduce(g) / torch.distributed.get_world_size()
        Memory_Bank.v.data = args.mem_momentum * Memory_Bank.v.data + g #+ args.mem_wd * Memory_Bank.W.data
        Memory_Bank.W.data = Memory_Bank.W.data - memory_lr * Memory_Bank.v.data
```

All other parts should be identical to MoCo's setting

**Comment by handsomelys** [comment]
Thank you for your kind reply!!! 
I will try to use CaCo for my tasks.
